### ISSUE TITLE
Refine `VersionManagerError` for More Detailed Error Handling

### ISSUE THREAD

**Issue by harshdoesdev** [shuru-project/shuru#4]
#### Background
In the current `Error` enum, the `VersionManagerError` variant is used as a catch-all for errors related to the version management process, but it only passes a generic string message. This makes it difficult to distinguish between different types of version management issues (e.g., download failures, extraction errors, build issues, etc.).

To improve error handling and provide more detailed error information, we should create more specific error types under the `VersionManagerError` variant. This will help pinpoint the exact nature of the failure and improve the clarity of the error messages.

#### Task
1. **Create Specific Error Types**:
   - Introduce more granular error types within the `VersionManagerError` to handle specific scenarios such as:
     - Download failure
     - Extraction failure
     - Build failure
     - Installation failure

2. **Update Error Enum**:
   - Replace the current string-based `VersionManagerError` with a detailed enum that categorizes different error types. The enum should include fields for context (e.g., file paths, URLs) when relevant.

3. **Refactor Code**:
   - Refactor the parts of the code that currently pass string messages for `VersionManagerError` to use the newly created specific error types.

#### Example

Current Error Enum:
```rust
#[derive(Debug, Error)]
pub enum Error {
    // Other variants...
    #[error("Version manager error: '{0}'")]
    VersionManagerError(String),
    // ...
}
```

Updated Error Enum:
```rust
#[derive(Debug, Error)]
pub enum Error {
    // Other variants...
    #[error("Version manager error")]
    VersionManagerError(#[from] VersionManagerError),
    // ...
}

#[derive(Debug, Error)]
pub enum VersionManagerError {
    #[error("Failed to download version from {url}: {source}")]
    DownloadError {
        url: String,
        source: reqwest::Error,
    },
    #[error("Failed to extract archive: {path}")]
    ExtractionError {
        path: String,
        source: std::io::Error,
    },
    #[error("Build process failed.")]
    BuildError,
    #[error("Installation failed.")]
    InstallationError,
}
```

#### Steps to Complete:
1. **Define the `VersionManagerError` enum** in `src/error.rs` with appropriate variants for download, extraction, build, and installation failures.
2. **Refactor existing code** that uses `VersionManagerError(String)` to use the new detailed variants, such as in the version manager implementation for Node.js and Python.
3. **Update error handling logic** where relevant, to construct and return the new `VersionManagerError` types.
4. **Test the changes** by simulating different failure scenarios to ensure the correct error type and message are produced.

#### Difficulty
Beginner-friendly, but requires understanding of Rust's error handling patterns and enums.

#### Files to Modify
- `src/error.rs`
- Version manager code in `src/version_manager/*.rs`

This issue will improve error traceability and make debugging easier for users and developers.

**Comment by ChrisX101010** [comment]
@harshdoesdev 👍

**Comment by airvzxf** [comment]
Hi, @harshdoesdev.

I wonder if it is good to have this output format for this refactor: “_Error: General error | Specific error | Technical error_”. For example:
```text
Error: Version manager error | Failed to download version from www.fake.com | builder error
```

**Comment by prxjwal** [comment]
could you assign me this, thanks

**Comment by ChrisX101010** [comment]
Assign me, this is my part..

**Comment by ChrisX101010** [comment]
> could you assign me this, thanks

This is my part of the work, not yours..

**Comment by prxjwal** [comment]
Sorry for the confusion @ChrisX101010 I guess accidentally u have removed yourself from the assignment following that i requested to assign this task to me

**Comment by airvzxf** [comment]
Sorry, @ChrisX101010.

**Please, please, please, don't take my next words in a bad way,** it is based on my experience.

I am 7 years old of experience contributing to GitHub / Open Source.

In open source, these actions are common, the action to enter and work on an issue which is assigned because at the end it provides more opportunities to get a quality code; provide other points of view and more advantages. For example, if you create a pull request with your changes, them will be different from mine, the owner of the project could have the possibility to check what is the best idea or implementation for his project.

Regarding specific **Hacktoberfest**, usually, there are many people who ask for the assignation but never do the work. I created 3 reports last years for Hactoberfest and anyone contributed nothing. I don't say that you are not contributing, I am saying that it is a common pattern, the reasons could be vast: time, lack of knowledge, abandon, and more. The best strategic is doing the work, creating the pull request and send.

My advice for you is to upload your pull request, instead you didn't finish, as this way the owner of the project could take in account your effort. The risk is that others “stolen” your code and create a faster or better solution. But in the end, Open Source is not a competition, is an international contribution to the code, and these behaviors that I tell you is the best way to create open code with quality, and free.

**Comment by airvzxf** [comment]
Hi, @prxjwal.

Same advance for you, in the same way, provide better open-source culture. I recommend not to focus on request assignations, focus on solving the problem or improving the feature, and send you code (Pull Request).

**Comment by harshdoesdev** [comment]
Hi everyone, sorry I was busy due to personal reasons, but I will be active now. @ChrisX101010, I believe you may have accidentally unassigned yourself from this issue, which is why @prxjwal thought it was open for reassignment. As per your request, I am reassigning the issue to you. Please open a PR so that the progress can be tracked. Also, keep in mind that everyone here is helping to improve the project, so please take suggestions or comments as constructive feedback.

**Comment by harshdoesdev** [comment]
Hi @prxjwal, I’ll be opening some new Rust issues today, and we can discuss them. I’ll tag you there.

**Comment by harshdoesdev** [comment]
Thanks for the suggestions, @airvzxf. The error output format you suggested is much better. @ChrisX101010, I believe we can implement this in the current PR.

> Another question, what is the expected output when we have chained errors?
> In the below example, the format is “Error: General error | Specific error | General error | Specific error | Technical error”

In the case of chained errors, I really like how Rust handles them. Instead of putting everything on a single line, we can format the errors in a way that makes them more “human-readable,” for example:

```
Error: Version manager error
    Caused by: Failed to extract archive
        Caused by: Command execution error
            Caused by: Failed to open tar.gz file
                Caused by: Permission denied (os error 13)
```

Please share your thoughts on this approach.

**Comment by harshdoesdev** [comment]
Closed with: https://github.com/shuru-project/shuru/pull/24

**Comment by harshdoesdev** [comment]
Hi @ChrisX101010, the changes mentioned in this issue have been addressed in PR [#24](https://github.com/shuru-project/shuru/pull/24), which has already been merged. Therefore, we are closing this issue.
